pkg/cnsmigration/migrator.go

```diff
@@ -0,0 +1,315 @@
+			_, isAlreadyExistErr := soapFault.VimFault().(vim.AlreadyExists)
+			if isAlreadyExistErr {
+				// Volume already exists in the target SP, hence return success.
+				return nil
+			}
+		}
+		return err
+	}
+	taskInfo, err := task.WaitForResultEx(ctx)
+	if err != nil {
+		msg := fmt.Errorf("error waiting for relocation task: %v", err)
+		printErrorObject(msg)
+		return msg
+	}
+	results := taskInfo.Result.(types.CnsVolumeOperationBatchResult)
+	for _, result := range results.VolumeResults {
+		fault := result.GetCnsVolumeOperationResult().Fault
+		if fault != nil {
+			printError("Fault: %+v encountered while relocating volume %v", fault, volumeID)
+			return fmt.Errorf(fault.LocalizedMessage)
+		}
+	}
+	return nil
+}
+
+func (c *CNSVolumeMigrator) checkForDatastore(pv *v1.PersistentVolume) bool {
+	csiSource := pv.Spec.CSI
+	for _, cnsVolume := range c.matchingCnsVolumes {
+		vh := csiSource.VolumeHandle
+		if cnsVolume.VolumeId.Id == vh {
+			printGreenInfo("found a volume to migrate: %s", vh)
+			pvcName, podName, inUseFlag := c.usedVolumeCache.volumeInUse(volumeHandle(vh))
+			if inUseFlag {
+				c.failedToMigrate++
+				printError("volume %s is being used by pod %s in pvc %s", vh, podName, pvcName)
+				return false
+			}
+			return true
+		}
+	}
+	c.volumesNotFound++
+	printInfo("Unable to find volume %s in CNS datastore %s", pv.Name, c.sourceDatastore)
+	return false
+}
+
+func (c *CNSVolumeMigrator) getCNSVolumes(ctx context.Context, dsName string) error {
+	ds, err := c.getDatastore(ctx, dsName)
+	if err != nil {
+		return err
+	}
+
+	cnsQueryFilter := types.CnsQueryFilter{
+		Datastores: []vim.ManagedObjectReference{ds.Reference()},
+	}
+	for {
+		res, err := c.vSphereConnection.CnsClient().QueryVolume(ctx, cnsQueryFilter)
+		if err != nil {
+			return err
+		}
+		c.matchingCnsVolumes = append(c.matchingCnsVolumes, res.Volumes...)
+		if res.Cursor.Offset == res.Cursor.TotalRecords || len(res.Volumes) == 0 {
+			break
+		}
+		cnsQueryFilter.Cursor = &res.Cursor
+	}
+	return nil
+}
+
+func (c *CNSVolumeMigrator) getDatastore(ctx context.Context, dsName string) (*object.Datastore, error) {
+	finder := find.NewFinder(c.vSphereConnection.VimClient(), false)
+	dc, err := finder.Datacenter(ctx, c.vSphereConnection.DefaultDatacenter())
+	if err != nil {
+		return nil, fmt.Errorf("can't find datacenters %s", c.vSphereConnection.DefaultDatacenter())
+	}
+
+	finder.SetDatacenter(dc)
+	ds, err := finder.Datastore(ctx, dsName)
+	if err != nil {
+		return nil, fmt.Errorf("error finding datastore %s in datacenter %s", dsName, c.vSphereConnection.DefaultDatacenter())
+	}
+	return ds, nil
+}
+
+func (c *CNSVolumeMigrator) loginToVCenter(ctx context.Context) error {
+	infra, err := c.openshiftConfigClientSet.ConfigV1().Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+	if err != nil {
+		printError("error getting infrastrucure object: %v", infra)
+		return err
+	}
+
+	klog.V(3).Infof("Creating vSphere connection")
```

Comment:
The rest looks like `createVCenterConnection()`, but I did not check thoroughly.

Reply:
That code is evolving, I want to keep it separate for now.